### DISCUSSION TITLE
Parallel execution and @BeforeAll

### DISCUSSION THREAD

**Discussion by vagra348** [cucumber/common#2205] [Q&A Java]
Hello! I'm working on the project where are thousands of UI-tests, so it's necessary for us to make them running in parallel. 
Also there is cookie "token" that refresh every time user logs into the website. 

I don't want to hard-code token values for every user every day before Jenkins starts working, so I need to write an authorisation-by-token scenarios for some users with REST-requests and need to run this feature Before All of the other features. 

Somewhere I read BeforeAll tag will run once per worker if parallel mode is used, so I suppose tokens will be rewritten at the same moment as many times as we have threads..... looks like a horrible mess 

So (if it is possible) how can I solve this problem with Cucumber? Thanks for reading

**Comment by mpkorstanje** [comment]
Cucumber JVM only executes the before all and after all hooks once per execution.

> Somewhere I read BeforeAll tag will run once per worker if parallel mode is used, so I suggest tokens will be rewritten at the same moment as many times as we have threads..... looks like a horrible mess

Could you clarify where you read that? Then we can fix the documentation.

**Reply by vitalets** [reply]
Maybe it's from CucumberJS [docs](https://github.com/cucumber/cucumber-js/blob/main/docs/parallel.md#hooks):

> When using parallel mode, any BeforeAll and AfterAll hooks you have defined will run once per worker.

**Reply by vagra348** [reply]
Yes, this is exactly what I've read before!
But now I see it's for JS...

**Reply by vagra348** [reply]
> Cucumber JVM only executes the before all and after all hooks once per execution.
>
So how will look like the execution? 
Firstly the BeforeAll (calling my Auth.feature or something like that) will be run in the main thread (and other N threads will be sleeping while that) and only after that method all features will be executed?

**Reply by mpkorstanje** [reply]
I think it is best if you make a small project to try things out with first. You'll get a better understanding that way.

**Reply by vagra348** [reply]
Hello again! Today I understand it a bit better (tried just to print something to console with before and after annotations), but now I have another related problem - is there any method or command or property/option for Runner class to execute a special feature? 
Our runner class looks like:

imports:...
@RunWith(Cucumber.class)
@CucumberOptions(monochrome = true, plugin = {"pretty"}, glue = {"..."}, features = {"..."}, tags = {"@all"})
public class CucumberTest { }

I want to use @BeforeClass (our project is built on cucumber 4.x.x) in this class and call here only one feature with its special tag or name or something else
I found in google only 'options-tags' feature calling and I fell into Cucumber.class (I saw getChildren(), runChild() methods that work with FeatureRunner) but feel now I'm reading the wrong info and moving away from right action

Is there an ability to make such script:
public class CucumberTest {
    @BeforeClass
    ***calling Auth.feature*** so it will run separately before others
}
Please help 🙏

**Reply by mpkorstanje** [reply]
> I want to use @BeforeClass (our project is built on cucumber 4.x.x)

I'm not sure if those work in with version of Cucumber. The last v4.x version was released 4 years ago. You should consider upgrading to the latest version of Cucumber. You can find more information in the [release-notes](https://github.com/cucumber/cucumber-jvm/tree/main/release-notes).

>  in this class and call here only one feature with its special tag or name or something else

You can not do that with Cucumber. Though there shouldn't be anything that prevents you from having multiple runner classes, each targeting a different set of features and then ordering the execution of those runner classes with Surefire. Though that all gets hacky, fragile and complicated quickly. It is also not recommended to share state between Scenarios (or tests in general) as it makes them brittle and hard to debug.

Consider instead obtaining the authentication code in a purely programmatic fashion. If done lazily and with caching it might even be possible avoid the need for hooks al together.
